docs/experimental/nnx/nnx_basics.md

```diff
@@ -0,0 +1,285 @@
+
+```{code-cell} ipython3
+# Module replacement
+pretrained = Block(dim=2, rngs=nnx.Rngs(42)) # imagine this is pretrained
+model.blocks[0] = pretrained
+# Module sharing
+model.blocks[3] = model.blocks[1]
+# Monkey patching
+def awesome_layer(x): return x
+model.blocks[2] = awesome_layer
+
+# Variable sharing (weight tying)
+model.blocks[-1].linear.variables.kernel = model.blocks[0].linear.variables.kernel
+
+model(jnp.ones((1, 2)))
+```
+
+## The Functional API
+
+The Functional API established a clear boundary between reference/object semantics and 
+value/pytree semantics. It also allows same amount of fine-grained control over the 
+state Linen/Haiku users are used to. The Functional API consists of 3 basic methods: 
+`split`, `merge`, and `update`.
+
+The `StatefulLinear` Module shown below will serve as an example to learn to use the 
+Functional API. It contains some `nnx.Param` Variables and a custom `Count` Variable
+type which is used to keep track of integer scalar state that increases on every 
+forward pass.
+
+```{code-cell} ipython3
+class Count(nnx.Variable): pass
+
+class StatefulLinear(nnx.Module):
+  def __init__(self, din: int, dout: int, *, rngs: nnx.Rngs):
+    self.w = nnx.Param(jax.random.uniform(rngs(), (din, dout)))
+    self.b = nnx.Param(jnp.zeros((dout,)))
+    self.count = Count(0)
+
+  def __call__(self, x: jax.Array):
+    self.count += 1
+    return x @ self.w + self.b
+  
+model = StatefulLinear(din=2, dout=3, rngs=nnx.Rngs(0))
+```
+
+### State and GraphDef
+
+A Module can be decomposed into a `State` and `GraphDef` pytrees using the 
+`.split()` method. State is a Mapping from strings to Variables or nested 
+States. GraphDef is contains all the static information needed to reconstruct 
+a Module graph, its analogous to JAX's `PyTreeDef`, and for convenience it 
+implements an empty pytree.
+
+```{code-cell} ipython3
+state, static = model.split()
+
+print(f'{state = }\n')
+print(f'{static = }'[:200] + '...')
+```
+
+### Split, Merge, and Update
+
+`merge` is the reverse of `split`, it takes the GraphDef + State and reconstructs
+the Module. As shown in the example below, by using split and merge in sequence 
+any Module can be lifted to be used in any JAX transform. `update` can
+update a Module strucure from a compatible State, this is often used to propagate the state
+updates from a transform back to the source object outside.
+
+```{code-cell} ipython3
+print(f'{model.count = }')
+
+# 1. Use split to create a pytree representation of the Module
+state, static = model.split()
+
+@jax.jit
+def forward(static: nnx.GraphDef, state: nnx.State, x: jax.Array):
+  # 2. Use merge to create a new model inside the JAX transformation
+  model = static.merge(state)
+  # 3. Call the Module
+  y = model(x)
+  # 4. Use split to propagate State updates
+  state, _ = model.split()
+  return y, state
+
+y, state = forward(static, state, x=jnp.ones((1, 2)))
+# 5. Update the state of the original Module
+model.update(state)
```

Comment:
How is `model.update(state)` different from `static.merge(state)`?

Reply:
`update` mutable the original references with the new state, `merge` creates new references with the given state.